File: 198.打家劫舍.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(!n)
            return 0;
        if(n == 1)
            return nums[0];
        int *money = new int[n];
        money[0] = nums[0];
        money[1] = nums[1];
        int max = money[0] > money[1] ? money[0] : money[1];
        for(int i=2; i<n; ++i){
            money[i] = 0;
            for(int j=i-2; j>=0; --j){
                if(money[i] < nums[i] + money[j])
                    money[i] = nums[i] + money[j];
            }
            if(money[i] > max)
                max = money[i];
        }
        return max;
    }

File: 198.打家劫舍.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var rob = function(nums) {
    let n = nums.length;
    if (n == 0)
        return 0;
    if (n == 1)
        return nums[0];
    // dp[i]表示前i个房子可偷窃的最大金额
    let dp = new Array(n).fill(0);
    dp[0] = nums[0];
    dp[1] = Math.max(nums[0], nums[1]);
    for (let i = 2; i < n; ++i) {
        dp[i] = Math.max(dp[i - 2] + nums[i], dp[i - 1]);
    }
    return dp[n - 1];
};